<a href="https://colab.research.google.com/github/MarioViniciusBC/regressao_logistica/blob/main/Regressao_logistica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Regressão Logística

A regressão logística é um dos algoritmos mais fundamentais e amplamente utilizados no aprendizado de máquina, projetado especificamente para resolver problemas de classificação. Apesar de carregar a palavra "regressão" no nome — uma herança de suas raízes estatísticas —, seu objetivo prático não é prever valores contínuos, mas sim categorizar informações, frequentemente de forma binária (como aprovar/rejeitar, fraude/legítimo, spam/caixa de entrada). Ela realiza essa tarefa calculando a probabilidade de um evento ocorrer e estabelecendo um limite de decisão claro. Por ser rápida, computacionalmente eficiente e altamente interpretável, a regressão logística é, na grande maioria das vezes, o modelo base inicial (baseline) escolhido por cientistas de dados antes de partirem para abordagens mais complexas.

### **Como a Regressão Logística Aprende**

Entender como a regressão logística aprende é como acompanhar o treinamento de um detetive que, no início, tenta adivinhar o resultado, mas vai refinando suas habilidades a cada erro cometido.

#### **1. Inicialização e o Primeiro Palpite**
O processo começa de forma essencialmente cega, através da **Inicialização**. O modelo atribui valores iniciais e aleatórios, chamados de **pesos** ($w$), para cada característica dos dados de entrada ($x$), além de um valor de ajuste básico chamado **viés** ($b$).

Com esses números em mãos, ele cria uma combinação linear simples, calculando uma reta matemática expressa por:

$$z = w_1x_1 + w_2x_2 + \dots + b$$

Esse resultado numérico é então "injetado" na **função sigmoide**, cuja fórmula é:

$$\sigma(z) = \frac{1}{1 + e^{-z}}$$



O que sai dessa equação matemática é o primeiro "chute" do modelo: uma probabilidade bruta entre **0 e 1**, que o detetive usa para classificar o dado.

#### **2. Avaliando o Erro: A Função de Custo**
Como esse primeiro palpite foi baseado em pesos aleatórios, ele muito provavelmente estará bastante errado. Para que o modelo perceba a distância entre o seu palpite e a realidade, ele precisa ser avaliado por uma **Função de Custo** (Cost Function), que na regressão logística é chamada de **Log Loss** (ou entropia cruzada).

Essa métrica atua como um professor implacável:
* Se o modelo previu que a chance de um e-mail ser spam era de **99%** e ele realmente era, a penalidade calculada é mínima.
* No entanto, se o modelo calculou **95%** de certeza de que era spam, mas a mensagem era do seu chefe, o Log Loss aplica uma penalidade matemática severa.

O objetivo central de todo o treinamento passa a ser descobrir como reduzir esse custo (erro) o máximo possível.

#### **3. Corrigindo a Rota: Gradiente Descendente**
Para corrigir a rota e diminuir esse erro gigantesco, o modelo aciona o seu principal mecanismo de ajuste contínuo: o algoritmo de **Gradiente Descendente**.

Imagine o modelo de olhos vendados no topo de um vale muito irregular (onde o topo representa o erro máximo) precisando encontrar o ponto mais profundo e baixo (onde o erro é quase zero).



Usando o cálculo das derivadas, o algoritmo mede a "inclinação" do erro em relação a cada peso atual. Sentindo para qual lado o terreno desce, ele ajusta ligeiramente os valores originais dos pesos ($w$) e do viés ($b$). A velocidade com que ele altera esses valores é controlada pela **taxa de aprendizado** (learning rate), que dita se ele dará passos cautelosos ou saltos largos ladeira abaixo.

#### **4. Iteração e Convergência**
Por fim, a inteligência do modelo é consolidada através da **Iteração** e **Convergência**. O aprendizado não acontece em uma única tentativa; o modelo precisa revisar os dados inúmeras vezes.

Ele repete o ciclo inteiro por milhares de vezes:
1. Gerar a previsão usando a sigmoide.
2. Medir o estrago com o Log Loss.
3. Atualizar os pesos com o Gradiente Descendente.

Cada passagem completa por todos os dados de treinamento é chamada de **época** (epoch). A cada época, a curva em formato de "S" se ajusta com mais precisão para separar corretamente os dados. O treinamento só chega ao fim quando as atualizações não conseguem mais reduzir o erro de forma perceptível. É nesse fundo do vale que dizemos que o modelo **convergiu**, os pesos foram otimizados e ele está pronto para fazer previsões no mundo real.

#### **5. A Magia por Trás: A Transformação Logit**
A transformação **logit** é exatamente a ponte matemática inteligente que conecta o mundo das probabilidades (que vivem espremidas entre 0 e 1) com o mundo das equações lineares infinitas que o modelo calcula.

Lembra que no início do aprendizado o modelo cria aquela reta simples, expressa por $z = w_1x_1 + \dots + b$? O problema é que uma reta pode crescer para o infinito positivo ou negativo, enquanto a probabilidade jamais pode passar de 1 ou cair abaixo de 0.

A transformação logit surge para resolver essa incompatibilidade geométrica: ela é, na verdade, a função inversa da curva sigmoide. Em vez de transformar um número infinito em uma probabilidade, o logit pega uma probabilidade e a "estica" de volta para o infinito.

#### **Entendendo a "Chance" (Odds)**
Para entender como ele faz isso, precisamos primeiro compreender o conceito de **chance** (odds), que é simplesmente a probabilidade de um evento acontecer dividida pela probabilidade de ele não acontecer:

$$\text{Chance} = \frac{p}{1 - p}$$

* **Exemplo:** Se a probabilidade ($p$) de um e-mail ser spam é de **0,8** (80%), a probabilidade de não ser é de **0,2** (20%). A "chance" é de 0,8 dividido por 0,2, resultando em 4. Dizemos que a chance é de **4 para 1**.

Transformar probabilidades em chances já é um grande avanço, porque o valor pode crescer até o infinito positivo, mas ainda esbarra no limite do zero (não existem chances negativas).

#### **O Logaritmo Natural**
É aqui que acontece a mágica final da transformação logit: aplicamos o **logaritmo natural** em cima dessas chances. A fórmula completa do logit se torna:

$$\text{logit}(p) = \ln\left(\frac{p}{1-p}\right)$$

O logaritmo transforma valores fracionários (entre 0 e 1) em números negativos que vão até o infinito, e valores maiores que 1 em números positivos que também vão até o infinito. Ao calcular o logaritmo natural das chances, criamos uma escala de valores que espelha perfeitamente a nossa equação linear (de $-\infty$ a $+\infty$).

#### **Conclusão**
Portanto, quando dizemos que estamos treinando uma "regressão logística", o que o modelo está realmente fazendo por trás dos panos é uma **regressão linear comum**. Porém, em vez de a reta tentar prever a probabilidade diretamente (o que quebraria a matemática), a reta está prevendo o logit da probabilidade.

A verdadeira equação fundamental do modelo é:

$$\ln\left(\frac{p}{1-p}\right) = w_1x_1 + w_2x_2 + \dots + b$$

Essa transformação é o que permite que o algoritmo use a matemática simples e eficiente das linhas retas para resolver problemas complexos de classificação em curva.

## Exemplo da Base de Crédito

# Exemplo Prático: Avaliação de Risco de Crédito

## Enunciado do Problema
Uma instituição financeira precisa automatizar a análise de concessão de empréstimos. Para isso, ela possui uma base de dados histórica (`risco_credito.pkl`) contendo o perfil de diversos clientes e a classificação final do risco que eles representaram. O objetivo é treinar um modelo de Regressão Logística capaz de analisar o perfil de novos clientes (histórico de crédito, dívida, garantias e renda) e prever se o risco de conceder o empréstimo será **alto** ou **baixo**.

*Nota: A base original contém alguns registros classificados como risco "moderado". Como queremos focar na regressão logística binária (duas classes), precisaremos limpar esses dados antes do treinamento.*

In [13]:
import numpy as np
from sklearn.linear_model import LogisticRegression

### Base de dados: Risco de crédito

In [5]:
import pickle

with open('risco_credito.pkl', 'rb') as f:
  X_risco_credito, y_risco_credito = pickle.load(f)

In [6]:
X_risco_credito

array([[2, 0, 1, 0],
       [1, 0, 1, 1],
       [1, 1, 1, 1],
       [1, 1, 1, 2],
       [1, 1, 1, 2],
       [1, 1, 0, 2],
       [2, 1, 1, 0],
       [2, 1, 0, 2],
       [0, 1, 1, 2],
       [0, 0, 0, 2],
       [0, 0, 1, 0],
       [0, 0, 1, 1],
       [0, 0, 1, 2],
       [2, 0, 1, 1]], dtype=object)

In [7]:
y_risco_credito # Temos que apagar os registros 2, 7, 11

array(['alto', 'alto', 'moderado', 'alto', 'baixo', 'baixo', 'alto',
       'moderado', 'baixo', 'baixo', 'alto', 'moderado', 'baixo', 'alto'],
      dtype=object)

In [10]:
X_risco_credito = np.delete(X_risco_credito, [2, 7, 11], axis = 0)
y_risco_credito = np.delete(y_risco_credito, [2, 7, 11], axis = 0)

In [11]:
X_risco_credito

array([[2, 0, 1, 0],
       [1, 0, 1, 1],
       [1, 1, 1, 2],
       [1, 1, 1, 2],
       [1, 1, 0, 2],
       [2, 1, 1, 0],
       [0, 1, 1, 2],
       [0, 0, 0, 2],
       [0, 0, 1, 0],
       [0, 0, 1, 2],
       [2, 0, 1, 1]], dtype=object)

In [12]:
y_risco_credito

array(['alto', 'alto', 'alto', 'baixo', 'baixo', 'alto', 'baixo', 'baixo',
       'alto', 'baixo', 'alto'], dtype=object)

In [14]:
logistic_risco_credito = LogisticRegression(random_state = 1)
logistic_risco_credito.fit(X_risco_credito, y_risco_credito)

LogisticRegression(random_state=1)

In [15]:
logistic_risco_credito.intercept_

array([-0.80855152])

In [16]:
logistic_risco_credito.coef_

array([[-0.76706797,  0.23921064, -0.47989768,  1.12196145]])

### Previsões

In [18]:
# história boa, dívida alta, garantia nenhuma, renda > 35
# história ruim, dívida_alta, garantias adequada, renda < 15
previsao_1 = logistic_risco_credito.predict([[0, 0, 1, 2], [2, 0, 0, 0]])
previsao_1

array(['baixo', 'alto'], dtype=object)

## Explicação Passo a Passo

### Passo 1: Importação de Bibliotecas e Dados
Começamos importando o `numpy` para manipulação de matrizes e o `LogisticRegression` da biblioteca `scikit-learn`. Em seguida, usamos a biblioteca `pickle` para abrir e carregar nossa base de dados pré-processada, dividindo-a em duas variáveis:
* **$X$ (Atributos previsores):** As características do cliente (matriz com os dados).
* **$y$ (Classe alvo):** A resposta histórica se o risco foi alto, moderado ou baixo.

### Passo 2: Limpeza e Preparação (Binarizando o Problema)
Ao inspecionar a variável alvo ($y$), notamos a presença da classe "moderado" nos índices 2, 7 e 11. Para que nosso modelo trabalhe com uma classificação estritamente binária (apenas "alto" ou "baixo"), utilizamos a função `np.delete` para remover essas três linhas tanto da matriz de atributos ($X$) quanto das respostas ($y$).

### Passo 3: Treinamento do Modelo
Instanciamos o nosso modelo chamando `LogisticRegression(random_state = 1)`. O parâmetro `random_state` garante que a inicialização aleatória dos pesos seja sempre a mesma, permitindo que os resultados sejam reproduzíveis.

Em seguida, executamos o comando `fit(X, y)`. É neste momento exato que o algoritmo de Gradiente Descendente entra em ação, rodando as equações matemáticas para encontrar os melhores pesos para a nossa base.



### Passo 4: Extraindo a Equação da Reta
Após o treinamento, podemos "espiar" o que o modelo aprendeu acessando seus parâmetros internos:
* **`intercept_`**: Representa o valor do viés ($b$), que neste caso foi ajustado para aproximadamente `-0.808`.
* **`coef_`**: Representa os pesos matemáticos ($w$) atribuídos a cada uma das 4 características do cliente. O modelo gerou a matriz `[-0.767, 0.239, -0.479, 1.121]`.

Isso significa que a equação linear interna do nosso modelo ficou configurada da seguinte forma:
$$z = -0.767x_1 + 0.239x_2 - 0.479x_3 + 1.121x_4 - 0.808$$

### Passo 5: Fazendo Previsões no Mundo Real
Com o modelo treinado, usamos a função `predict` para avaliar dois novos clientes hipotéticos que acabaram de chegar no banco:

**Cliente 1: `[0, 0, 1, 2]`**
* História boa, Dívida alta, Nenhuma garantia, Renda > 35.
* **Previsão do modelo:** Risco `baixo`.

**Cliente 2: `[2, 0, 0, 0]`**
* História ruim, Dívida alta, Garantia adequada, Renda < 15.
* **Previsão do modelo:** Risco `alto`.

O modelo aplicou os pesos matemáticos nas características de cada cliente, passou o resultado pela função sigmoide e entregou a classificação final com base na probabilidade calculada!